File: leetcode/98.py
# 98. Validate Binary Search Tree

class Solution:
    def isValidBST(self, root: Optional[TreeNode], left = float('-inf'), right = float('inf')) -> bool:
        if not root:
            return True
        if root.val <= left or root.val >= right:
            return False
        return self.isValidBST(root.left, left, min(right, root.val)) and self.isValidBST(root.right, max(root.val, left), right)

'''
Runtime: 44 ms, faster than 72.15% of Python3 online submissions for Validate Binary Search Tree.
Memory Usage: 16.4 MB, less than 79.87% of Python3 online submissions for Validate Binary Search Tree.
'''

File: leetcode/98.js
// 98. Validate Binary Search Tree

var isValidBST = function(root, left=Number.NEGATIVE_INFINITY, right=Number.POSITIVE_INFINITY) {
    if (!root) {
        return true;
    } 
    if (root.val <= left || root.val >= right) {
        return false;
    }
    return isValidBST(root.left, left, Math.min(right, root.val)) && isValidBST(root.right, Math.max(left, root.val), right)
};

/*
Runtime: 68 ms, faster than 99.52% of JavaScript online submissions for Validate Binary Search Tree.
Memory Usage: 43.4 MB, less than 30.77% of JavaScript online submissions for Validate Binary Search Tree.
*/